Sources/SwiftSyntax/RawSyntax.swift

```diff
@@ -10,1096 +10,291 @@
+  /// The syntax kind of this raw syntax.
+  var kind: SyntaxKind {
+    switch rawData.payload {
+    case .materializedToken(_): return .token
+    case .layout(let dat): return dat.kind
     }
-    return raw
   }
 
-  /// Create a token node using the programmatic APIs.
-  static func create(
-    kind: TokenKind,
-    leadingTrivia: Trivia,
-    trailingTrivia: Trivia,
-    length: SourceLength,
-    presence: SourcePresence
-  ) -> RawSyntax {
-    let tokdata = ConstructedTokenData(kind: kind, leadingTrivia: leadingTrivia,
-      trailingTrivia: trailingTrivia)
-    let (data, capacity) = TokenData.dataAndExtraCapacity(for: tokdata)
-    let buffer = self.create(minimumCapacity: capacity) { _ in
-      RawSyntaxBase(data: data, byteLength: length.utf8Length,
-        isPresent: presence == .present)
-    }
-    let raw = unsafeDowncast(buffer, to: RawSyntax.self)
-    raw.withUnsafeMutablePointerToElements {
-      TokenData.initializeExtra(tokdata, extraPtr: $0)
-    }
-    return raw
+  /// Whether or not this node is a token one.
+  var isToken: Bool {
+    kind == .token
   }
 
-  deinit {
-    return withUnsafeMutablePointers {
-      $0.pointee.deinitialize(extraPtr: $1)
-    }
+  /// Whether or not this node is a collection one.
+  var isCollection: Bool {
+    kind.isSyntaxCollection
   }
 
-  /// The syntax kind of this raw syntax.
-  var kind: SyntaxKind {
-    return header.kind
+  /// Whether or not this node is an unknown one.
+  var isUnknown: Bool {
+    kind.isUnknown
   }
 
-  /// Whether or not this node is a token one.
-  var isToken: Bool {
-    return header.isToken
+  /// Child nodes.
+  var children: RawSyntaxBuffer {
+    switch rawData.payload {
+    case .materializedToken(_): return .init(start: nil, count: 0)
+    case .layout(let dat): return dat.layout
+    }
   }
 
   func child(at index: Int) -> RawSyntax? {
-    return withUnsafeMutablePointers {
-      $0.pointee.child(at: index, extraPtr: $1)
-    }
+    guard hasChild(at: index) else { return nil }
+    return children[index]
   }
 
   func hasChild(at index: Int) -> Bool {
-    return withUnsafeMutablePointers {
-      $0.pointee.hasChild(at: index, extraPtr: $1)
-    }
+    children[index] != nil
   }
 
   /// The number of children, `present` or `missing`, in this node.
   var numberOfChildren: Int {
-    return header.numberOfChildren
+    return children.count
   }
 
   /// Total number of nodes in this sub-tree, including `self` node.
   var totalNodes: Int {
-    return header.totalNodes
+    switch rawData.payload {
+    case .materializedToken(_): return 1
+    case .layout(let dat): return dat.descendantCount + 1
+    }
   }
 
   var presence: SourcePresence {
-    return header.isPresent ? .present : .missing
+    let isMissing = (self.byteLength == 0 &&
+                     !self.isCollection &&
+                     !self.isUnknown &&
+                     !(isToken && self.rawTokenKind == .eof))
+    return isMissing ? .missing : .present
```

Comment:
Maybe it’s just me but the logic wasn’t immediately clear to me. Suggestion:

```suggestion
    if self.byteLength != 0 {
      // The node has source text associated with it. It's present.
      return .present
    }
    if self.isCollection || self.isUnknown {
      // We always consider collections 'present' because they can just be empty.
      return .present
    }
    if isToken && self.rawTokenKind == .eof {
      // The end of file token never has source code associated with it but we
      // still consider it valid.
      return .present
    }

    // If none of the above apply, the node is missing.
    return .missing
```